cmd/main/main.go

```diff
@@ -113,132 +90,22 @@ func main() {
-		if fgaReplicaClientConn != nil {
-			defer fgaReplicaClientConn.Close()
-		}
-	}
-
-	aclClient := acl.NewACLClient(fgaClient, fgaReplicaClient, redisClient)
-
-	// Create tls based credential.
-	var creds credentials.TransportCredentials
-	var tlsConfig *tls.Config
-	if config.Config.Server.HTTPS.Cert != "" && config.Config.Server.HTTPS.Key != "" {
-		tlsConfig = &tls.Config{
-			ClientAuth: tls.RequireAndVerifyClientCert,
-			MinVersion: tls.VersionTLS12,
-		}
-		creds, err = credentials.NewServerTLSFromFile(config.Config.Server.HTTPS.Cert, config.Config.Server.HTTPS.Key)
-		if err != nil {
-			logger.Fatal(fmt.Sprintf("failed to create credentials: %v", err))
-		}
-		grpcServerOpts = append(grpcServerOpts, grpc.Creds(creds))
-	}
-
-	grpcServerOpts = append(grpcServerOpts, grpc.MaxRecvMsgSize(constant.MaxPayloadSize))
-	grpcServerOpts = append(grpcServerOpts, grpc.MaxSendMsgSize(constant.MaxPayloadSize))
-
-	pipelinePublicServiceClient, pipelinePublicServiceClientConn := external.InitPipelinePublicServiceClient(ctx)
-	if pipelinePublicServiceClientConn != nil {
-		defer pipelinePublicServiceClientConn.Close()
-	}
-	mgmtPublicServiceClient, mgmtPublicServiceClientConn := external.InitMgmtPublicServiceClient(ctx)
-	if mgmtPublicServiceClientConn != nil {
-		defer mgmtPublicServiceClientConn.Close()
-	}
-	mgmtPrivateServiceClient, mgmtPrivateServiceClientConn := external.InitMgmtPrivateServiceClient(ctx)
-	if mgmtPrivateServiceClientConn != nil {
-		defer mgmtPrivateServiceClientConn.Close()
-	}
-
-	// Initialize MinIO client
-	minIOParams := minio.ClientParams{
-		Config: config.Config.Minio,
-		Logger: logger,
-		AppInfo: minio.AppInfo{
-			Name:    serviceName,
-			Version: serviceVersion,
-		},
-	}
-	minIOFileGetter, err := minio.NewFileGetter(minIOParams)
-	if err != nil {
-		logger.Fatal("Failed to create MinIO file getter", zap.Error(err))
-	}
+	// Get gRPC server options and credentials
+	grpcServerOpts, creds, tlsConfig := newGrpcOptionAndCreds(logger)
 
-	retentionHandler := service.NewRetentionHandler()
-	minIOParams.ExpiryRules = retentionHandler.ListExpiryRules()
-	minIOClient, err := minio.NewMinIOClientAndInitBucket(ctx, minIOParams)
-	if err != nil {
-		logger.Fatal("failed to create MinIO client", zap.Error(err))
-	}
+	privateGrpcS := grpc.NewServer(grpcServerOpts...)
+	reflection.Register(privateGrpcS)
 
-	artifactPublicServiceClient, artifactPublicServiceClientConn := external.InitArtifactPublicServiceClient(ctx)
-	if artifactPublicServiceClientConn != nil {
-		defer artifactPublicServiceClientConn.Close()
-	}
+	publicGrpcS := grpc.NewServer(grpcServerOpts...)
+	reflection.Register(publicGrpcS)
 
-	artifactPrivateServiceClient, artifactPrivateServiceClientConn := external.InitArtifactPrivateServiceClient(ctx)
-	if artifactPrivateServiceClientConn != nil {
-		defer artifactPrivateServiceClientConn.Close()
-	}
+	// Initialize all clients
+	pipelinePublicServiceClient, mgmtPublicServiceClient, mgmtPrivateServiceClient,
```

Comment:
I don't particularly love this way of initialising the clients. I've recently needed to modify the output of this script a couple of times (in a service with the similar pattern) and I had to kind of fish for the client I wanted to change. The same problem but for long argument lists is explained [here](https://google.github.io/styleguide/go/best-practices#function-argument-lists) quite clearly.

I saw nothing wrong in initialising each client in the main function but if we want to extract that logic to make it more readable and compact, I'd suggest returning a `clients` type (which can have a `close` method itself) to make the results more accessible.